docs/proposals/memory-management.md

```diff
@@ -0,0 +1,135 @@
+
+####Issues with Current model
+The life of a node is blissful up until it has memory pressure, at which points the kernel resorts to killing processes.
+
+* Memory is inherently overcommitted.
+
+ This combined with lack of swap space, increases the probability of system OOMs. This slows down the responsiveness of the system and could have secondary effects like high tail latencies for system daemons.
+
+* A container in a pod with Restart Always policy will be restarted continuously even if it OOMs. 
+
+* Other than the Guaranteed class, jobs in other classes don't have a memory limit.
+
+* Any process in a container can be killed due to either system memory pressure or exceeding container limits.
+
+ There is no container cleanup policy in the event of an OOM kill. What should happen to a pod if a process in one of its container's was OOM killed? 
+
+* Kubelet cannot associate a system memory pressure OOM kill to a specific container or Pod. 
+
+* Memory usage of system daemons is unrestricted. This results in dynamic memory availability on the node. 
+
+* All burstable jobs that exceed their request is equally likely to be killed. Ideally, the job that exceeds the most should be killed. 
+
+* There are no limits placed on kernel resources consumed by jobs (proc, fd, sockets, network buffers, etc)
+
+#####Ideas for improving memory management
+
+1. Define OOM handling policy for containers and Pods:
+
+    *Suggestion*: Whenever an OOM kill occurs in a container, mark that entire container as OOM killed and terminate it. We can provide a knob to alter this behavior if required
+    If a Pod has an app container in "guaranteed" class and a "side-car" container in "Burstable" class, what happens to the pod when the "side-car" is OOM killed? Will the Pod continue to run without the "side-car"?
+    Requiring all containers in a Pod to be of the same QoS class will be simpler.
+
+2. Impose limits on "best effort" and "Burstable" jobs
+
+   *Suggestion*: Don't overcommit memory assigned to "guaranteed" tasks.
+   Use the remaining memory for "Burstable" and best-effort jobs. Whenever there is memory pressure in the "Burstable" + "best effort" bucket, a "best effort" task will most likely be killed based on our QoS policy. 
+
+3. Restrict resource usage of system daemons
+
+   Allocate only the remaining memory to jobs (reserve some memory for the kernel).
+   Even if this might be wasteful, this results in reliable performance. Memory leaks in system daemons are supposed to be caught and fixed before being deployed in production.
+
+4. Set soft limits for Burstable jobs to their memory request: This should result in "Burstable" jobs being pushed to their requested memory limit under system memory pressure. 
+
+5. Kubelet reduces the likelihood of system memory pressure:
+
+  1. Kubelet can subscribe to memory pressure notifications from the kernel, and possibly evict low priority jobs (or)
```

Comment:
the doc uses "jobs" a lot. But since we haven't figured out how we want processes, containers, and pods to interact long-term with respect to cgroup hierarchies, restart policy, etc., I'd argue a misnomer like "jobs" is probably the best thing to do -- keep the distinction ambiguous until we figure out exactly what policy we want